### PR TITLE
[WIP] Support topk classification metrics [ci skip]

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Add support for topk in classification metrics. Let's see how it goes with tests.
TODO:
- [ ] Add more tests
- [ ] Update class metrics

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by oke-aditya** [comment]
This looks better and cleaner. I guess natively supporting `top_k` is far more useful.

File: pytorch_lightning/metrics/functional/classification.py [hunk 1]
```diff
@@ -134,10 +134,38 @@ def stat_scores(
     return tp, fp, tn, fn, sup
 
 
+def _get_topk(
+        pred: torch.Tensor,
+        target: torch.Tensor,
+        topk: Optional[int] = None,
+) -> torch.Tensor:
+    if (target.ndim > 1) and (topk is not None):
+        raise ValueError(
```
Comment (justusschock) [inline pytorch_lightning/metrics/functional/classification.py:143]: Can you add a test case for this?

File: pytorch_lightning/metrics/functional/classification.py [hunk 2]
```diff
@@ -134,10 +134,38 @@ def stat_scores(
...
+        )
+
+    if (target.ndim > 1) or (pred.ndim == 1):
+        return pred
+
+    if topk is None:
+        topk = 1
+
+    if topk > pred.size(1):
+        raise ValueError(
```
Comment (justusschock) [inline pytorch_lightning/metrics/functional/classification.py:154]: also a test_case for this :)

File: pytorch_lightning/metrics/functional/classification.py [hunk 3]
```diff
@@ -134,10 +134,38 @@ def stat_scores(
...
+        pred: torch.Tensor,
+        target: torch.Tensor,
+        topk: Optional[int] = None,
+) -> torch.Tensor:
+    if (target.ndim > 1) and (topk is not None):
+        raise ValueError(
+            f'topk={topk} is not supported for multi-label target values.'
+        )
+
+    if (target.ndim > 1) or (pred.ndim == 1):
```
Comment (justusschock) [inline pytorch_lightning/metrics/functional/classification.py:147]: I don't fully understand this case... Can you explain, why to return the whole pred if `target.ndim>1`?

File: pytorch_lightning/metrics/functional/classification.py [hunk 4]
```diff
@@ -241,16 +269,18 @@ def accuracy(
...
 ) -> torch.Tensor:
     """
     Computes the accuracy classification score
 
     Args:
-        pred: predicted labels
+        pred: predicted labels or probabilities
         target: ground truth labels
         num_classes: number of classes
+        topk: number of most likely outcomes considered to find the correct label
```
Comment (justusschock) [inline pytorch_lightning/metrics/functional/classification.py:283]:
```suggestion
        topk: number of most likely outcomes considered to find the correct label. Defaults to the usual top1 accuracy
```

File: pytorch_lightning/metrics/functional/classification.py [hunk 5]
```diff
@@ -341,9 +377,10 @@ def precision_recall(
     Computes precision and recall for different thresholds
 
     Args:
-        pred: estimated probabilities
+        pred: predicted labels or probabilities
         target: ground-truth labels
         num_classes: number of classes
+        topk: number of most likely outcomes considered to find the correct label
```
Comment (justusschock) [inline pytorch_lightning/metrics/functional/classification.py:383]:
```suggestion
        topk: number of most likely outcomes considered to find the correct label. Defaults to the usual top1.
```

File: pytorch_lightning/metrics/functional/classification.py [hunk 6]
```diff
@@ -381,15 +423,17 @@ def precision(
...
 ) -> torch.Tensor:
     """
     Computes precision score.
 
     Args:
-        pred: estimated probabilities
+        pred: predicted labels or probabilities
         target: ground-truth labels
         num_classes: number of classes
+        topk: number of most likely outcomes considered to find the correct label
```
Comment (justusschock) [inline pytorch_lightning/metrics/functional/classification.py:436]:
```suggestion
        topk: number of most likely outcomes considered to find the correct label. Defaults to the usual top1.
```

File: pytorch_lightning/metrics/functional/classification.py [hunk 7]
```diff
@@ -408,23 +452,25 @@ def precision(
...
 ) -> torch.Tensor:
     """
     Computes recall score.
 
     Args:
-        pred: estimated probabilities
+        pred: predicted labels or probabilities
         target: ground-truth labels
         num_classes: number of classes
+        topk: number of most likely outcomes considered to find the correct label
```
Comment (justusschock) [inline pytorch_lightning/metrics/functional/classification.py:473]:
```suggestion
        topk: number of most likely outcomes considered to find the correct label. Defaults to the usual top1.
```

**Review by justusschock — CHANGES_REQUESTED** [review]

File: pytorch_lightning/metrics/functional/classification.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/metrics/functional/classification.py:143]: yeah sure, it's in TODO actually.

File: pytorch_lightning/metrics/functional/classification.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/metrics/functional/classification.py:154]: yeah sure.

File: pytorch_lightning/metrics/functional/classification.py (hunk 3 above)
Comment (rohitgr7) [reply pytorch_lightning/metrics/functional/classification.py:147]: multi-label classification with `topk=None`. `topk` doesn't work for multi-label classification right?

**Comment by teddykoker** [comment]
Here is a top k accuracy implementation using the new `Metrics` API if it's helpful:
```python
class TopKAccuracy(pl.metrics.Metric):
    def __init__(self, k=[1, 5], dist_sync_on_step=False):
        super().__init__(dist_sync_on_step=dist_sync_on_step)
        self.k = k
        self.add_state("correct", default=torch.zeros(len(k)), dist_reduce_fx="sum")
        self.add_state("total", default=torch.tensor(0.0), dist_reduce_fx="sum")

    def update(self, logits, y):
        _, pred = logits.topk(max(self.k), dim=1)
        pred = pred.t()
        corr = pred.eq(y.view(1, -1).expand_as(pred))
        self.correct += torch.stack([corr[:k].view(-1).sum() for k in self.k])
        self.total += y.numel()

    def compute(self):
        return self.correct.float() / self.total
```

**Comment by oke-aditya** [comment]
This is correct. I used the same formula earlier. Can we simply keep `topk` as `int` instead of tuple ?

**Comment by teddykoker** [comment]
@oke-aditya I think I prefer that as well, especially since `self.log(metric)` only supports scalar values

**Comment by oke-aditya** [comment]
I guess we can replace `.view()` by `.reshape()`. It recently threw an error on torch 1.7. It didn't throw in 1.6 though.
Here is the modfied code.

```
class AccuracyTopK(pl.metrics.Metric):
    def __init__(self, top_k=1, dist_sync_on_step=False):
        super().__init__(dist_sync_on_step=dist_sync_on_step)
        self.k = top_k
        self.add_state("correct", default=torch.tensor(0.0), dist_reduce_fx="sum")
        self.add_state("total", default=torch.tensor(0.0), dist_reduce_fx="sum")

    def update(self, logits, y):
        _, pred = logits.topk(self.k, dim=1, True, True)
        pred = pred.t()
        corr = pred.eq(y.reshape(1, -1).expand_as(pred))
        self.correct += corr[:self.k].reshape(-1).float().sum()
        self.total += y.numel()

    def compute(self):
        return self.correct.float() / self.total
```

**Comment by rohitgr7** [comment]
@oke-aditya thanks. I'll complete this one in a few days. The metrics package has been changed and I haven't checked it out yet.

**Comment by oke-aditya** [comment]
Great. 😀

**Comment by zhutmost** [comment]
Looks that AccuracyTopK is not available in the current version? Will it be in the next version?

**Comment by rohitgr7** [comment]
Hi @zhutmost it will be added in https://github.com/PyTorchLightning/pytorch-lightning/pull/4838

will close this one after https://github.com/PyTorchLightning/pytorch-lightning/pull/4838 gets merged. Although you can use the implementation of AccuracyTopK from above for now.

**Comment by Borda** [comment]
@rohitgr7 @justusschock how is it going here, ready to review/merge? 🐰

**Comment by rohitgr7** [comment]
closing this, will be added in #4837 and follow up PRs.
